### ISSUE TITLE
Trained models

### ISSUE THREAD

**Issue by kargarisaac** [cirpue49/kitti_pretrained_pp#1]
Hi,
What is the trained model for? cars or people or what? In the paper, they mentioned that they trained 2 models for cars and pedestrians/cyclists? 
If you use one model for all classes, what is the parameter values? such as thresholds, etc?
Thank you

**Comment by lucasjinreal** [comment]
@kargarisaac  SECOND official repo has merged point pillars,  I have trained a light version on all classes, it seems very good

**Comment by kargarisaac** [comment]
@jinfagang 
Thank you for your response. I want to train pointpillars for multiclass object detection, but the results are almost 0 for all classes. Is it possible to share your config file for multiclass?

**Comment by josephshaaban** [comment]
Hello @kargarisaac @jinfagang ,
I want to train official SECOND with PointPillars configuration on my own data. My own data is a point cloud and 3D bounding boxes with labels.
I want to know if SECOND supports this dataset structure. I have a deadline and I have to present a proof of concept, Can you help me with your answers?
Thank you!

**Comment by kargarisaac** [comment]
@josephshaaban 
Hi,
I didn't get any good answer yet. the results are all zero for me. I mixed the config file of car and ped_cycle but no result yet

**Comment by lucasjinreal** [comment]
here is a show demo models trained on kitti

![](https://camo.githubusercontent.com/17b80b0e45c33fc45214f112543ebbd74a650135/68747470733a2f2f73322e617831782e636f6d2f323031392f30342f32352f455a676143442e676966)

**Comment by kargarisaac** [comment]
@jinfagang 
perfect. Is it possible to share the config file?

**Comment by jeannotes** [comment]
@jinfagang are you training with SECOND or pointpillow? really appreciate if you share the model

**Comment by lucasjinreal** [comment]
@kargarisaac @jeannotes I am simply using official repo config file. you can got same result after training. I am not using pointpillars, VoxelLite version. I am currently training on nuScenes. If you want track my status and share our results, you can post a topic at here: http://talk.strangeai.pro I will reply your guys

**Comment by jeannotes** [comment]
how do you get these "pre.onnx" as well as "rpn.onnx"?
@kargarisaac @josephshaaban @jeannotes

**Comment by spchuang** [comment]
Also would like to know how pre.onnx and rpn.onnx are generated. I took a stab at trying to export ONNX from SECOND repo but the param and model layers seem to be quite different

**Comment by lucasjinreal** [comment]
@spchuang It can not be done, SECOND uses lots of self-defined layers and it's spconv lib were used as python wrapper. Hope if there any one can export it into onnx since SECOND it's the most accurate and realtime 3D detection methods through.

**Comment by spchuang** [comment]
In that case, do you know how the ONNX files here for autoware's pointpillar model are generated? They seem to have a custom python implementation then.

**Comment by kargarisaac** [comment]
@spchuang 
No I don't know how to do that. autoware guys do not answer

**Comment by jjarquin** [comment]
I dont know how the Autoware guys exported the models, but I've found a trick that lets you export as a onnx file using the SECOND point pillars implementation. The main problems are the dictionaries that are used on pytorch, the opset versions, and some operations that are not supported by the onnx to tensor rt conversor.

It is possible though, having trained the model, to load it, modify the forward function and then export the model as onnx.

For the RPN network simply modify the forward function inside the rpn.py file
The modified forward function looks like this. 

forward method of the class RPNBase
```
    def forward(self, x):
        res = super().forward(x)
        x = res["out"]
        box_preds = self.conv_box(x)
        cls_preds = self.conv_cls(x)

        box_preds = box_preds.permute(0, 2, 3, 1).contiguous()
        cls_preds = cls_preds.permute(0, 2, 3, 1).contiguous()
        conv_dir_cls = self.conv_dir_cls(x)
        conv_dir_cls = conv_dir_cls.permute(0, 2, 3, 1).contiguous()

        return (box_preds, cls_preds, conv_dir_cls)
```
And an example of how to export it to onnx
[export_rpn.txt](https://github.com/k0suke-murakami/kitti_pretrained_point_pillars/files/3407021/export_rpn.txt)

**Comment by SmallMunich** [comment]
> I dont know how the Autoware guys exported the models, but I've found a trick that lets you export as a onnx file using the SECOND point pillars implementation. The main problems are the dictionaries that are used on pytorch, the opset versions, and some operations that are not supported by the onnx to tensor rt conversor.
> 
> It is possible though, having trained the model, to load it, modify the forward function and then export the model as onnx.
> 
> For the RPN network simply modify the forward function inside the rpn.py file
> The modified forward function looks like this.
> 
> forward method of the class RPNBase
> 
> ```
>     def forward(self, x):
>         res = super().forward(x)
>         x = res["out"]
>         box_preds = self.conv_box(x)
>         cls_preds = self.conv_cls(x)
> 
>         box_preds = box_preds.permute(0, 2, 3, 1).contiguous()
>         cls_preds = cls_preds.permute(0, 2, 3, 1).contiguous()
>         conv_dir_cls = self.conv_dir_cls(x)
>         conv_dir_cls = conv_dir_cls.permute(0, 2, 3, 1).contiguous()
> 
>         return (box_preds, cls_preds, conv_dir_cls)
> ```
> 
> And an example of how to export it to onnx
> [export_rpn.txt](https://github.com/k0suke-murakami/kitti_pretrained_point_pillars/files/3407021/export_rpn.txt)

Hi, @jjarquin  I see this export_rpn.txt, line 39 this np.newaxis, can you explain the structure simplely ? And, this can support multi-class convertion, I have some errors to debug...

**Comment by jjarquin** [comment]
Hi @SmallMunich , I actually collected most of the code from the point pillars source code in the second repository. The np.newaxis is is simply adding a new dimension at the beginning, and comes in handy for example to add a new dimension that represents the batches that will be fed to the net (I guess that is the reason of this here).  
This is for point pillars trained for single class, I haven't done any multi class my self so I cannot help you much there.

**Comment by SmallMunich** [comment]
> Hi @SmallMunich , I actually collected most of the code from the point pillars source code in the second repository. The np.newaxis is is simply adding a new dimension at the beginning, and comes in handy for example to add a new dimension that represents the batches that will be fed to the net (I guess that is the reason of this here).
> This is for point pillars trained for single class, I haven't done any multi class my self so I cannot help you much there.

@jjarquin ok, thanks for you reply.

**Comment by SmallMunich** [comment]
> Hi @SmallMunich , I actually collected most of the code from the point pillars source code in the second repository. The np.newaxis is is simply adding a new dimension at the beginning, and comes in handy for example to add a new dimension that represents the batches that will be fed to the net (I guess that is the reason of this here).
> This is for point pillars trained for single class, I haven't done any multi class my self so I cannot help you much there.

@jjarquin ok, I get it. thanks for your reply.  best wishes.

**Comment by dotorl127** [comment]
> @josephshaaban
> Hi,
> I didn't get any good answer yet. the results are all zero for me. I mixed the config file of car and ped_cycle but no result yet

hi, you mentioned training with custom config file as mixed default config files. right?
how was your work result?

**Comment by asimonov-hitachi** [comment]
hi everyone, I believe this is how the model was saved:

https://github.com/k0suke-murakami/train_point_pillars/blob/master/second/pytorch/train.py#L896

did anyone figure out what config was used for training?
